File: practice_java_program/src/array_java/FindMissingNumber.java
package array_java;

//11. Find the Missing Number:
//Given an array containing n distinct numbers taken from the range 0 to n, find the one missing from the array.
//Calculate the expected sum of the first n natural numbers using the formula n * (n + 1) / 2
//Calculate the actual sum of the elements in the array using a ‘for’ loop.
//Subtract the actual sum from the expected sum.
//The result is the missing number.
//Option 1: Static Method Approach

public class FindMissingNumber {
    
    // Method to find missing number
    public static int findMissingNumber(int[] nums) {
        int n = nums.length; // For array containing numbers 0 to n, length should be n
        int expectedSum = n * (n + 1) / 2;
        int actualSum = 0;
        
        for (int num : nums) {
            actualSum += num;
        }
        return expectedSum - actualSum;
    }
    
    public static void main(String[] args) {
        // Test cases
        int[] testArray1 = {0, 1, 2, 4, 5}; // Missing 3
        int[] testArray2 = {3, 0, 1}; // Missing 2
        int[] testArray3 = {0, 1, 2, 3, 4, 5, 6, 8, 9}; // Missing 7
        
        System.out.println("Missing number in array 1: " + findMissingNumber(testArray1));
        System.out.println("Missing number in array 2: " + findMissingNumber(testArray2));
        System.out.println("Missing number in array 3: " + findMissingNumber(testArray3));
    }
}

////Option 2: Instance Method Approach
//
//public class FindMissingNumber {
//    
//    // Instance method
//    public int findMissingNumber(int[] nums) {
//        int n = nums.length;
//        int expectedSum = n * (n + 1) / 2;
//        int actualSum = 0;
//        
//        for (int num : nums) {
//            actualSum += num;
//        }
//        return expectedSum - actualSum;
//    }
//    
//    public static void main(String[] args) {
//        FindMissingNumber finder = new FindMissingNumber();
//        
//        // Test cases
//        int[] testArray1 = {0, 1, 2, 4, 5}; // Missing 3
//        int[] testArray2 = {3, 0, 1}; // Missing 2
//        int[] testArray3 = {0, 1, 2, 3, 4, 5, 6, 8, 9}; // Missing 7
//        
//        System.out.println("Missing number in array 1: " + finder.findMissingNumber(testArray1));
//        System.out.println("Missing number in array 2: " + finder.findMissingNumber(testArray2));
//        System.out.println("Missing number in array 3: " + finder.findMissingNumber(testArray3));
//    }
//}

File: practice_python_program/program/list_array_programs/find_missing_number.py
"""
Interview Question:
Given an array containing n distinct numbers taken from the range 1 to n+1, find the one missing number.

Interview Explanation:
In automation testing (e.g., verifying sequential transaction IDs, sequence numbering, or order IDs in database records), identifying missing sequence gaps is a standard validation task.

Key Concepts:
1. Mathematical formula method: Sum of first N numbers = N * (N + 1) // 2.
2. XOR method: Avoiding integer overflow issues.
3. Time Complexity: O(n), Space Complexity: O(1).
"""


def find_missing_number_sum(nums: list) -> int:
    """
    Finds missing number in 1 to N+1 sequence using mathematical sum formula.
    """
    n = len(nums) + 1
    expected_sum = n * (n + 1) // 2
    actual_sum = sum(nums)
    return expected_sum - actual_sum


def find_missing_number_xor(nums: list) -> int:
    """
    Finds missing number using bitwise XOR operator (handles large numbers safely).
    """
    n = len(nums) + 1
    xor_all = 0
    for i in range(1, n + 1):
        xor_all ^= i

    for num in nums:
        xor_all ^= num

    return xor_all


if __name__ == "__main__":
    sample_sequence = [1, 2, 4, 5, 6, 7, 8]  # 3 is missing

    missing_sum = find_missing_number_sum(sample_sequence)
    missing_xor = find_missing_number_xor(sample_sequence)

    print(f"Sequence: {sample_sequence}")
    print(f"Missing Number (Sum Method): {missing_sum}")
    print(f"Missing Number (XOR Method): {missing_xor}")
